pyoomph/c_output.py

```diff
@@ -22,224 +22,186 @@
-            statement = self.emit_statement(ast.statement).rstrip().rstrip(";")
-            expression = self.emit_expression(ast.expression)
-            return f"(({statement}), ({expression}))"
-        raise NotImplementedError(ast)
-
-    def emit_label(self, name: str) -> str:
-        # It's invalid c syntax to end a block with a label, (void)0 fixes
-        return f"{name}: (void)0;\n\t"
-
-    def emit_statement(self, ast: tast.Statement) -> str:
-        if isinstance(ast, tast.CreateLocalVar):
-            self.add_local_var(ast.var)
-            return f"{self.variable_names[ast.var]} = {self.emit_expression(ast.value)};\n\t"
-
-        if isinstance(ast, tast.SetLocalVar):
-            return f"{self.variable_names[ast.var]} = {self.emit_expression(ast.value)};\n\t"
-
-        if isinstance(ast, (tast.ReturningCall, tast.VoidCall)):
-            return self.emit_call(ast) + ";\n\t"
-
-        if isinstance(ast, tast.DecRef):
-            return self.file_emitter.emit_decref(
-                self.emit_expression(ast.value), ast.value.type
+        if isinstance(ins, tast.CallConstructor):
+            return self.emit_call(
+                "ctor_" + self.file_emitter.get_type_c_name(ins.result.type),
+                ins.args,
+                ins.result,
             )
-
-        if isinstance(ast, tast.Return):
-            if ast.value is not None:
-                return f"""
-                retval = {self.emit_expression(ast.value)};
-                {self.file_emitter.emit_incref("retval", ast.value.type)}
-                goto out;
-                """
+        if isinstance(ins, tast.Return):
+            if ins.value is not None:
+                return f"{self.incref_var(ins.value)}; retval = {self.emit_local_var(ins.value)}; goto out;"
             return "goto out;"
-
-        if isinstance(ast, tast.If):
+        if isinstance(ins, tast.GetAttribute):
+            return f"{self.emit_local_var(ins.result)} = {self.emit_local_var(ins.obj)}->memb_{ins.attribute}; {self.incref_var(ins.result)};"
+        if isinstance(ins, tast.PointersEqual):
+            return f"{self.emit_local_var(ins.result)} = ({self.emit_local_var(ins.lhs)} == {self.emit_local_var(ins.rhs)});"
+        if isinstance(ins, tast.If):
+            then = "\n\t\t".join(map(self.emit_instruction, ins.then))
+            otherwise = "\n\t\t".join(map(self.emit_instruction, ins.otherwise))
             return f"""
-            if ({self.emit_expression(ast.condition)}) {{
-                {"".join(self.emit_statement(s) for s in ast.then)}
+            if ({self.emit_local_var(ins.condition)}) {{
+                {then}
             }} else {{
-                {"".join(self.emit_statement(s) for s in ast.otherwise)}
+                {otherwise}
             }}
             """
-
-        if isinstance(ast, tast.Loop):
+        if isinstance(ins, tast.Loop):
             # While loop because I couldn't get C's for loop to work here
```

Comment:
Comment outdated, now there are other (less subtle) reasons why this can't be C for loop:

```suggestion
```